packages/coreutils/src/instancetracker.ts

```diff
@@ -0,0 +1,423 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { CommandRegistry } from '@phosphor/commands';
+
+import {
+  PromiseDelegate,
+  ReadonlyJSONObject,
+  ReadonlyJSONValue
+} from '@phosphor/coreutils';
+
+import { IObservableDisposable } from '@phosphor/disposable';
+
+import { AttachedProperty } from '@phosphor/properties';
+
+import { ISignal, Signal } from '@phosphor/signaling';
+
+import { IDataConnector, IInstanceTracker } from './interfaces';
+
+/**
+ * A class that keeps track of widget instances on an Application shell.
+ *
+ * @typeparam T - The type of object being tracked.
+ */
+export class InstanceTracker<
+  T extends IObservableDisposable = IObservableDisposable
+> implements IInstanceTracker<T> {
+  /**
+   * Create a new instance tracker.
+   *
+   * @param options - The instantiation options for an instance tracker.
+   */
+  constructor(options: InstanceTracker.IOptions) {
+    this.namespace = options.namespace;
+  }
+
+  /**
+   * A signal emitted when an object instance is added.
+   *
+   * #### Notes
+   * This signal will only fire when an instance is added to the tracker.
+   * It will not fire if an instance injected into the tracker.
+   */
+  get added(): ISignal<this, T> {
+    return this._added;
+  }
+
+  /**
+   * A namespace for all tracked instances.
+   */
+  readonly namespace: string;
+
+  /**
+   * The current object instance.
+   *
+   * #### Notes
+   * The instance tracker does not set `current`. It is intended for client use.
```

Comment:
Interesting. I now see the difference in `currentWidget` vs `current`, hence the non-inheritance.